Remora.Plugins/Services/PluginService.cs

```diff
@@ -1,283 +1,258 @@
+        {
+            // Unload old plugin (file changed).
+            UnloadPlugin(Path.GetFileNameWithoutExtension(e.FullPath));
+
+            // Load new one.
+            LoadPlugin(e.FullPath, out var result);
+            if (!result.IsSuccess)
+            {
+                _options.ErrorDelegate?.Invoke(result);
+            }
+        };
+        _fileSystemWatcher.Created += (_, e) =>
+        {
+            // Load Plugin (file created).
+            LoadPlugin(e.FullPath, out var result);
+            if (!result.IsSuccess)
+            {
+                _options.ErrorDelegate?.Invoke(result);
+            }
+        };
+        _fileSystemWatcher.Deleted += (_, e) =>
+        {
+            // Unload plugin (file deleted).
+            UnloadPlugin(Path.GetFileNameWithoutExtension(e.FullPath));
+        };
+        _fileSystemWatcher.Renamed += (_, e) =>
+        {
+            // Unload old plugin (file renamed).
+            UnloadPlugin(Path.GetFileNameWithoutExtension(e.OldFullPath));
+
+            // Load new one.
+            LoadPlugin(e.FullPath, out var result);
+            if (!result.IsSuccess)
+            {
+                _options.ErrorDelegate?.Invoke(result);
+            }
+        };
+    }
+
+    /// <summary>
+    /// Loads all available plugins with a specific filter and watches them for any changes.
+    /// </summary>
+    public void LoadPlugins()
+    {
+        var assemblyPaths = GetPluginAssemblyPaths();
+        foreach (var assemblyPath in assemblyPaths)
+        {
+            LoadPlugin(assemblyPath, out var result);
+            if (!result.IsSuccess)
+            {
+                _options.ErrorDelegate?.Invoke(result);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Gets the search paths to plugin assemblies.
+    /// </summary>
+    /// <returns>
+    /// The search paths to plugin assemblies.
+    /// </returns>
+    /// <exception cref="InvalidOperationException">
+    /// When the parent directory is null.
+    /// </exception>
+    [Pure]
+    private IEnumerable<string> GetPluginAssemblyPaths()
+    {
+        var searchPaths = new List<string>();
+        if (_options.ScanAssemblyDirectory)
+        {
+            searchPaths.Add(GetApplicationDirectory());
+        }
+
+        searchPaths.AddRange(_options.PluginSearchPaths);
+        return searchPaths.Select
+        (
+            searchPath => Directory.EnumerateFiles
+            (
+                searchPath,
+                "*.dll",
+                SearchOption.AllDirectories
+            )
+        ).SelectMany(a => a);
+    }
+
+    /// <summary>
+    /// Loads a specific plugin and it's services.
+    /// </summary>
+    /// <param name="assemblyPath">
+    /// The path to the plugin assembly to load.
+    /// </param>
+    /// <param name="result">The result of loading the plugin.</param>
+    private void LoadPlugin(string assemblyPath, out Result result)

```

Comment:
Somewhere in my bot it crashes inside of here. I might have overlooked something here.